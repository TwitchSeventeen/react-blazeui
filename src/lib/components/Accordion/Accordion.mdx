---
name: Accordion
route: /components/accordion
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Accordion from './index';

# Accordion
<Playground>
    {() => {
        class TestAccordion extends React.Component {
          constructor(props) {
            super(props);
            this.state = { activeItem: 0 };
            this.handleClick = this.handleClick.bind(this);
          }
        
          handleClick(event, titleProps) {
            const { index } = titleProps;
            const { activeItem } = this.state;
            const newIndex = activeItem === index ? -1 : index;
            this.setState({ activeItem: newIndex });
          }
        
          render() {
            const { activeItem } = this.state;
            return (
              <Accordion>
                <Accordion.Title active={activeItem === 0} index={0} onClick={this.handleClick}>
                    Heading one
                </Accordion.Title>
                <Accordion.Content>
                    This is the body
                </Accordion.Content>
        
                <Accordion.Title active={activeItem === 1} index={1} onClick={this.handleClick}>
                    Heading two
                </Accordion.Title>
                <Accordion.Content>
                    This is the second body
                </Accordion.Content>
              </Accordion>
            );
          }
        }
        return <TestAccordion />;
    }}
    
</Playground>

The example above includes the logic required to make the Accordion work.

## Properties
### Accordion
<PropsTable of={Accordion} />

### Accordion.Title
<PropsTable of={Accordion.Title} />

### Accordion.Content
<PropsTable of={Accordion.Content} />