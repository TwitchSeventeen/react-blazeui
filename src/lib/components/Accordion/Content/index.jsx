import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Card from '../../Card';

function Content(props) {
  const {
    children, className, as: Component, dataName, ...other
  } = props;
  const classes = classNames('c-card__item--pane', className);
  return (
    <Component
      {...other}
      className={classes}
      dataName={dataName}
    >
      { children }
    </Component>
  );
}

Content.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
  /**
   * Element type to render as
   */
  as: PropTypes.elementType,
};

Content.defaultProps = {
  children: null,
  className: null,
  dataName: 'accordion-content',
  as: Card.Item,
};

export default Content;
