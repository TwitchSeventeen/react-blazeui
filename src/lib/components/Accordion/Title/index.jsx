import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

class Title extends React.Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
  }

  /**
   * Pass props to onClick when button is clicked
   * @param event
   */
  handleClick(event) {
    const { onClick } = this.props;

    onClick(event, this.props);
  }

  render() {
    const {
      children, className, active, dataName, index, ...other
    } = this.props;
    const classes = classNames('c-card__control', className);
    return (
      <button
        {...other}
        type="button"
        aria-expanded={active.toString()}
        className={classes}
        onClick={this.handleClick}
        index={index}
        data-name={dataName}
      >
        {children}
      </button>
    );
  }
}

Title.propTypes = {
  /**
   * Whether this is the active item
   */
  active: PropTypes.bool,
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * The index of the title in the accordion
   */
  index: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  /**
   * Function called when button is clicked
   */
  onClick: PropTypes.func,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
};

Title.defaultProps = {
  children: null,
  className: null,
  active: false,
  index: null,
  onClick: null,
  dataName: 'accordion-title',
};

export default Title;
