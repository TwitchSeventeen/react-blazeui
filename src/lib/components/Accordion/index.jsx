import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Card from '../Card';
import Title from './Title';
import Content from './Content';

function Accordion(props) {
  const { children, className, dataName, ...other } = props;
  const classes = classNames('c-card--accordion', className);
  return (
    <Card
      {...other}
      className={classes}
      data-name={dataName}
    >
      { children }
    </Card>
  );
}

Accordion.propTypes = {
  /**
   * This element can have children
   */
  children: PropTypes.node,
  /**
   * Additional classes
   */
  className: PropTypes.string,
  /**
   * Optional data-name attribute
   * (use for testing/selection)
   */
  dataName: PropTypes.string,
};

Accordion.defaultProps = {
  children: null,
  className: null,
  dataName: 'accordion',
};

Accordion.Title = Title;
Accordion.Content = Content;

export default Accordion;
