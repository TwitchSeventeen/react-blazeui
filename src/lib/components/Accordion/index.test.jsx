import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Accordion from './index';

describe('Accordion', () => {
  it('should be defined', () => {
    expect(Accordion).toBeDefined();
  });

  it('renders correctly', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
        <Accordion>
          lorem ipsum dolor
        </Accordion>
    );

    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });
});

