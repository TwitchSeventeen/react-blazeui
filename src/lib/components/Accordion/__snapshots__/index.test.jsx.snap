// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders correctly 1`] = `
<Card
  className="c-card--accordion"
  grouped={false}
>
  lorem ipsum dolor
</Card>
`;
